fairpy/items/saban_sethuraman_exchange_algorithm.py

```diff
@@ -0,0 +1,678 @@
+    >>> G = nx.to_dict_of_dicts(G, edge_data=None)
+    >>> stringify(G)
+    '{1:{a:{}}, 2:{a:{}}, 3:{a:{}}, 4:{a:{}}, a:{1:{}}, b:{2:{}}, c:{3:{}}, d:{4:{}}}'
+      """
+    # add edge between the agents to the houses they prefer.
+    graph = nx.DiGraph()
+    for i in PreferenceLists.keys():
+        for j in PreferenceLists[i][0]:
+            graph.add_edge(i, j)
+
+    # add edge between houses to the agent it belong to.
+    for i in owner_house.keys():
+        graph.add_edge(owner_house[i], i)
+    return graph
+
+
+def make_stack(graph, v, visited: list, stack: list):
+    visited.append(v)
+    for i in graph[v]:
+        if i not in visited:
+            make_stack(graph, i, visited, stack)
+    stack.append(v)
+
+
+def transpose_graph(graph):
+    """
+    Transpose the edges.
+    if G contains an edge (u, v) then the transpose of G contains an edge (v, u) .
+    :param graph:diGraph with agents and houses nodes and edges that connect between them.
+    :return:graph after transpose.
+
+    >>> graph = {'1': {'a': {}}, '2': {'b': {}}, 'c': {'3': {}}}
+    >>> graph = transpose_graph(graph)
+    >>> graph = nx.to_dict_of_dicts(graph, edge_data=None)
+    >>> stringify(graph)
+    '{1:{}, 2:{}, 3:{c:{}}, a:{1:{}}, b:{2:{}}, c:{}}'
+
+
+    >>> graph = {'1': {'a': {}, 'b' : {}}, '2': {'b': {}, 'a': {}}, 'c': {'3': {}}}
+    >>> graph = transpose_graph(graph)
+    >>> graph = nx.to_dict_of_dicts(graph, edge_data=None)
+    >>> stringify(graph)
+    '{1:{}, 2:{}, 3:{c:{}}, a:{1:{}, 2:{}}, b:{1:{}, 2:{}}, c:{}}'
+    """
+    TransposeG = nx.DiGraph()
+    for i in graph:
+        for j in graph[i]:
+            TransposeG.add_edge(j, i)
+    return TransposeG
+
+
+def SCC(graph, v, visited: list, SCC_list):
+    visited.append(v)
+    SCC_list.append(v)
+
+    for i in graph[v]:
+        if i not in visited:
+            SCC(graph, i, visited, SCC_list)
+
+    return SCC_list
+
+
+def if_all_satisfied(graph, owner_house: Dict):
+    """
+    O(n)
+    check if there aren't jealous agents in the graph return 1 else 0.
+    :param graph:diGraph with agents and houses nodes and edges that connect between them.
+    :param owner_house:dictionary that connect between the houses to there owners.keys = agent, values = houses.
+    :return:true if all the agents in the graph are satisfied, else false.
+
+    >>> PreferenceLists = {'1': [{'a', 'b', 'c'}],'2': [{'a', 'b', 'c'}],'3': [{'a', 'b', 'c'}]}
+    >>> house = {'1': 'a','2': 'b','3': 'c'}
+    >>> graph = make_graph_begin(PreferenceLists, house)
+    >>> if_all_satisfied(graph, house)
+    1
+
+    >>> PreferenceLists = {'1': [{'b'}],'2': [{'a', 'b', 'c'}],'3': [{'a', 'b', 'c'}]}
+    >>> house = {'1': 'a','2': 'b','3': 'c'}
+    >>> graph = make_graph_begin(PreferenceLists, house)
+    >>> if_all_satisfied(graph, house)
+    0
+    """
+    for li in graph:
+        if owner_house.keys().__contains__(li) and not graph.has_edge(li, owner_house[li]):
+            return 0
```

Comment:
הערכים הבוליאניים בפייתון הם לא 0,1 אלא True, False (מתחילים באות גדולה)

Reply:
I change it.